examples/async_fl/async_executor.py

```diff
@@ -63,8 +67,50 @@ def training_handler(self, clientId, conf, model=None):
 
         return train_res
 
+    def testing_handler(self, args, config=None):
+
+        evalStart = time.time()
+        device = self.device
+        model = config['test_model']
+        if self.task == 'rl':
+            client = RLClient(args)
+            test_res = client.test(args, self.this_rank, model, device=device)
+            _, _, _, testResults = test_res
+        else:
+            data_loader = select_dataset(self.this_rank, self.testing_sets,
+                                         batch_size=args.test_bsz, args=args,
+                                         isTest=True, collate_fn=self.collate_fn
+                                         )
+
+            if self.task == 'voice':
+                criterion = CTCLoss(reduction='mean').to(device=device)
+            else:
+                criterion = torch.nn.CrossEntropyLoss().to(device=device)
+
+            if self.args.engine == commons.PYTORCH:
+                test_res = test_model(self.this_rank, model, data_loader,
+                                      device=device, criterion=criterion, tokenizer=tokenizer)
+            else:
+                raise Exception(f"Need customized implementation for model testing in {self.args.engine} engine")
+
+            test_loss, acc, acc_5, testResults = test_res
+            logging.info("After aggregation round {}, CumulTime {}, eval_time {}, test_loss {}, test_accuracy {:.2f}%, test_5_accuracy {:.2f}% \n"
+                         .format(self.round, round(time.time() - self.start_run_time, 4), round(time.time() - evalStart, 4), test_loss, acc*100., acc_5*100.))
+
+        gc.collect()
+
+        return testResults
+
     def check_model_version(self, model_id):
-        return os.path.exists(self.temp_model_path_version(round))
+        return os.path.exists(self.temp_model_path_version(model_id))
+
+    def remove_stale_models(self, straggler_round):
```

Comment:
It seems a bit inefficient to check whether the path for all the models exist during each test handler call. Seems like it would be O(N^2) time?

Reply:
Cool! We can do it in the reversed order: 
```
while version(stale_model_id) exists:
        remove(stale_model_id)
        stale_model_id -= 1
```
On the other hand, since testing is not called in every iteration, this for loop is also fine to some extent. But we will definitely optimize it. Thanks for your input!